### DISCUSSION TITLE
Morse Runner design - Contest Class hierarchy

### DISCUSSION THREAD

**Discussion by w7sst** [w7sst/MorseRunner#81] [Ideas]
## Motivation
When I started adding multi-contest support to Morse Runner, it became very obvious that some refactoring would be needed. As the development continued, I started using lots of case statements in the code. These various case statements represent areas/features/behavior were a virtual function can be used to replace the case statement and move contest-specific code into the derived classes. I purposely implemented the prototype using these case statements to allow the coding patterns to appear in the code. To find examples in the code, search for `Adding a contest` in the root directory.

## Contest Class Hierarchy Proposal
This refactoring will include introducing a class hierarchy with a Contest base class and then have a derived Contest for each supported contest. Each derived contest will implement contest-specific rules and implement loading the associated Call History file. The diagram below shows a simple class hierarchy to show this concept.
``` mermaid
classDiagram
TMainForm *-- Contest
TMainForm: +TSimContest SimContest
TMainForm: +string Sent Exchange
TMainForm: +RecvExchTypes TExchTypes
TMainForm: +SetContest(TSimContest)
    Contest<|-- Wpx Contest
    Contest<|-- ARRL Field Day Contest
    Contest<|-- CWOPS CWT Contest
    Contest<|-- ARRL DX Contest
    Contest*-- MyStation
    Contest*-- CallListReader
    Contest: +LoadCallHistory(filename, user_callsign)
  Contest: +GetCallHistoryFileName() string
  Contest: +GetRecvExchTypes() TExchTypes
  Contest: +GetSentExchTypes() TExchTypes
  Contest: +SendMsg(station, TStationMessage)

class Station{
  +TExchTypes SentExchTypes
  +SendMsg(TStationMessage)
  +SendText(AMsg : string)
}

Station<|--MyStation
MyStation:   +SendText(AMsg string)

Station<|--DxStation

class CallListReader{
  +LoadCallHistory(filename, user_callsign)
  +PickCall() integer
  +GetCallInfo(integer) info
}
    class Wpx Contest{
        +GetCallHistoryFileName() string
        +SendMsg(station, TStationMessage)
    }
    class ARRL Field Day Contest{
        +GetCallHistoryFileName() string
        +SendMsg(station, TStationMessage)
    }
    class CWOPS CWT Contest{
        +GetCallHistoryFileName() string
        +SendMsg(station, TStationMessage)
    }
    class ARRL DX Contest{
        +GetCallHistoryFileName() string
         +GetRecvExchTypes() TExchTypes
         +GetSentExchTypes() TExchTypes
        +SendMsg(station, TStationMessage)
    }

CallListReader<|-- MasterDbaReader
CallListReader<|-- N1mmCallHistoryReader
N1mmCallHistoryReader<|-- WpxReader
N1mmCallHistoryReader<|-- CwopsCwtReader
N1mmCallHistoryReader<|-- ArrlFieldDayReader
CallListReader o-- "*" CallRecord

CallRecord: Callsign
CallRecord: Exch1
CallRecord: Exch2

class WpxReader{
  LoadCallHistory(filename, user_callsign)
  GetCall() string
}

class CwopsCwtReader{
  LoadCallHistory(filename, user_callsign)
  GetCall() string
}

class ArrlFieldDayReader{
  LoadCallHistory(filename, user_callsign)
  GetCall() string
}
```
> Note: The above class diagram is generated using Mermaid and is easily supported using markdown syntax. Visit [Creating Diagrams - GitHub docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) or [Include diagrams in your markdown files with Mermaid](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) for more information.

## Further Discussion
I would like to continue this discussion using the Comments below. As new topics are identified, we can create a separate Discussion topic as needed.

Please add some comments below and let me know what you are thinking about and any other areas you would like to see discussed.

**Comment by w7sst** [comment]
Hi @f6fvy @jr8ppg @jm1dxh @BH1SCW @tekenny @ct7aup @scotthibbs @n2ic
I would like your feedback on the following change...

## Introduce Dual Exchange Contest class
I'm planning on introducing an intermediate class, `TDualExchContest`, to represent common dual exchange DX Contest behaviors common to several DX contests. Examples include ARRL DX Contest, JA International Dx Contest, RSGB Contest, etc.

These contests have at least two common behaviors:
- local and dx stations may have different exchange types (see table below)
- a subset of the call history file is loaded (e.g. stations local to the contest region work DX stations)

Initially the `GetExchangeTypes()` function returns the exchange field types for Exchange 1 and exchange 2. For DX Contests, stations local to the contest send one exchange type while the DX Stations send a different exchange type as shown below. Note that RSGB DX Contest has the same exchange types for local and DX stations. A common implementation of `TDxContest.GetExchangeTypes` will handle all derived types.
Contest | Field Type 1 | Local station Field Type 2 | DX station Field Type 2
---------|--------------|-----------------------------|--------------------------
ARRL DX Contest | etRST | etStateProv | etPower
JA Intl DX Contest | etRST | etPrefectureNr | etCqZone
RSGB DX Contest | etRST | etSerialNr | etSerialNr

The `LoadCallHistory()` function will load the subset of the stations to be worked from the call history file. For example, in the ARRL DX Contest, a local US/CA station will work only DX stations. In this case, only DX stations callsign records will be loaded and returned during the contest simulation.

The class diagram below shows the proposed change. After discussion and code review, I will update the diagram at the top of this discussion.

``` mermaid
classDiagram
TMainForm *-- Contest
TMainForm: +TSimContest SimContest
TMainForm: +string Sent Exchange
TMainForm: +RecvExchTypes TExchTypes
TMainForm: +SetContest(TSimContest)
    Contest<|-- Wpx Contest
    Contest<|-- ARRL Field Day Contest
    Contest<|-- CWOPS CWT Contest
    Contest<|-- Dual Exch Contest
    Dual Exch Contest <|-- ARRL DX Contest
    Dual Exch Contest <|-- JA Intl DX Contest
    Dual Exch Contest <|-- RSGB DX Contest
    Contest*-- MyStation
  Contest: +OnSetMyCall()
  Contest: +OnContestPrepareToStart()
    Contest: +LoadCallHistory(user_callsign)
  Contest: +GetRecvExchTypes() TExchTypes
  Contest: +GetSentExchTypes() TExchTypes
  Contest: +GetExchangeTypes() TExchTypes

class Station{
  +TExchTypes SentExchTypes
}

Station<|--MyStation
Station<|--DxStation

class Dual Exch Contest {
  # TExchange2Type localExchType
  # TExchange2Type dxExchType
  # Construct(localSentExch, dxSentExch)
  +GetExchangeTypes() TExchTypes
}
class ARRL DX Contest {
+LoadCallHistory(user_callsign)
}
JA Intl DX Contest: +LoadCallHistory(user_callsign)
RSGB DX Contest: +LoadCallHistory(user_callsign)
```

**Comment by tekenny** [comment]
It seems good like a good structure to me.

**Comment by K6OK** [comment]
Is "DXContest" the right name for the class?  Later on MRCE might want to support some state QSO parties -- of course, those aren't DX contests but many of them have the same dual exchange structure of one exchange for in-state and another for out-of-state.  And CQ WW is a DX contest but it won't be using the DXContest class.  Maybe calling it ContestDualExchange might be more descriptive?

**Comment by w7sst** [comment]
Good point. Thank you for the input. Or DynamicExchangeContest.
